### DISCUSSION TITLE
AntennaPod sync?

### DISCUSSION THREAD

**Discussion by Gandalf-the-Blue** [SamTV12345/PodFetch#141] [Q&A] [answered]
### Describe the bug

Hi,

(Posted as an Issue and not a discussion, as I can't see the tab of Discussions on top)

I came here from the Reddit post you recently made on r/selfhosted. I was interested in setting up sync with the AntennaPod app. I've managed to self-host Podfetch and everything is working but how do I connect it to AntennaPod?

I tried to do the sync via the gpodder.net option where I provided my server url and username/password but nothing happened on the app - is this an issue due to Podfetch or AntennaPod?

I do see that you've mentioned RSS feeds as a way to get data into AntennaPod - but then how would the reverse work, i.e. if I listen to a podcast on antennapod how would it communicate that back to Podfetch?

### Reproduction

Docker + Antennapod

### System Info

```Shell
Docker deployed via Portainer on Debian based server behind a reverse proxy.
Working via browser.
```

### Used Package Manager

n/a

### Validations

- [X] Follow our [Code of Conduct](https://github.com/SamTV12345/.github/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guide](https://github.com/SamTV12345/.github/blob/main/CONTRIBUTING.md).
- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A, please open a GitHub Discussion instead.
- [X] The provided reproduction is a [minimal reproducible](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by SamTV12345** [comment]
Sorry forgot to setup discussions. You need to go to Settings->Synchronisation->Login and choose gpodder. After that enter the SERVER_URL that you used in your docker-compose configuration.

**Comment by SamTV12345** [comment, marked as answer]
From that on the paths deviate:
1. Basic auth: Enter your basic auth credentials
2. OIDC: Configure user via CLI

**Reply by Gandalf-the-Blue** [reply]
Hi,
I tried this. Added my server url -> https://podfetch.example.com
Added my basic auth details as Username & Password
There is a small loading wheel and then nothing happens.

**Reply by SamTV12345** [reply]
Can you try creating another user and login with that user?

**Reply by Gandalf-the-Blue** [reply]
That did it. So I guess the basic auth defined in the docker-compose doesn't actually create a user with that name?

Also, for anyone looking at this in the future
To create a user
Administration -> Invites -> '+' button -> copy invite link -> set up new account.

**Reply by Gandalf-the-Blue** [reply]
Hey @SamTV12345
Sorry, I guess I commented before checking everything. So now, AntennaPod says it is syncing everything but I don't see the AntennaPod data on podfetch + I added a podcast on podfetch and did a sync on Antennapod, but it doesn't appear in AntennaPod.

**Reply by SamTV12345** [reply]
Great that it worked. That is actually by design. It could be that you simply want to backup your AntennaPod data. Now you need to add the Podcasts you want to listen to in PodFetch. When a match is detected, e.g. the rss feeds are the same, you are able to listen to that podcast with every device.

**Reply by alexanderadam** [reply]
Hey @SamTV12345 I have just two tiny questions:

1. Would it in theory be possible to have an option that also adds new feeds to PodFetch if they were added via AntennaPod?
2. Is PodFetch keeping historical URLs and following redirects if there are some? So if a Podcast changes it's URL will PodFetch still be able to identify both feed URLs as the same podcast? Because this was one of the bugs with gPodder and AntennaPod won't keep a history on its own (see also [this issue](https://github.com/AntennaPod/AntennaPod/issues/2214) for a whole discussion on this topic). Apparently this merging is not trivial and needs to be done on the server side.

**Reply by SamTV12345** [reply]
1. How fine grain would be that be. Simply sending the remote podcast rss feed in PodFetch or do I need to tailor anything there? If yes I would need to download the podcast.
2. I didn't think about that yet. It's a good question on how to detect that this changed. If you change the complete url I have no idea what to do. How could I identify that simply the url changes?

**Reply by alexanderadam** [reply]
Thank you for the response @SamTV12345.

1. I don't know how this should work exactly. :grimacing: Maybe @keunes knows how the flow should be?
2. I created [an issue](https://github.com/SamTV12345/PodFetch/issues/175) for the duplication case.

**Reply by keunes** [reply]
> Would it in theory be possible to have an option that also adds new feeds to PodFetch if they were added via AntennaPod?

Currently? If gPodder.net connection/API is used, I think AntennaPod already sends out a signal, which PodFetch could pick up. It would work (more or less) the same [with the Open Podcast API](https://openpodcastapi.org/subscriptions/add-new.html) that we're currently making the specs for.

> It's a good question on how to detect that this changed. If you change the complete url I have no idea what to do. How could I identify that simply the url changes?

That's also a point that we've been discussing.
* If you get a sign from the client, it'd be fairly straightforward: update the URL for the subscription that was changed. I would imagine you get some form of ID from the client in order to identify the right subscription.
* If you notice the change yourself, e.g. if you pull a feed and get a ([301 or 308](https://github.com/AntennaPod/AntennaPod/wiki/Updated-feed-URLs-%28301-or-308-redirects%29)) redirect, you could recognise and record the change.

For syncing reliability, for the Open Podcast API we expect a server to track all past known URLs. This way, if a client joins the party and says 'Add this URL as subscription' but the URL is outdated, it won't lead to a duplicate subscription for the user.

Hope this helps. Happy to further discuss (although I don't know in detail how AP currently works).

**Reply by SamTV12345** [reply]
Thanks for the update. I'll check that out. As the download/sync part is already pretty complicated as there is a lot to track do you have a real world example where I can have the old podcast links and then redirect with 308 to get the new links?

**Reply by keunes** [reply]
> do you have a real world example where I can have the old podcast links and then redirect with 308 to get the new links?

I personally don't, but I believe @ByteHamster had a test set-up which he might be willing to share here.

**Reply by ByteHamster** [reply]
308: https://tools.bytehamster.com/podcast/redirect.php
`itunes:new-feed-url`: https://tools.bytehamster.com/podcast/redirect2.php

**Reply by SamTV12345** [reply]
Thanks. I don't know if I do something bad but I get always a 200. I did around 10 requests in Rust, switched to Postman did another 20 and still always get a 200. @ByteHamster

**Reply by SamTV12345** [reply]
You can check the code here https://github.com/SamTV12345/PodFetch/pull/177/files#diff-4d2793fdccdd3ca3c14fecb41dc8e1c6cee29834da0d40d3cd22ca05b2c65897R150

**Reply by alexanderadam** [reply]
@SamTV12345 it's definitely doing redirects.
You can test in in your browser, too.

`https://tools.bytehamster.com/podcast/redirect.php` will redirect to `https://tools.bytehamster.com/podcast/redirect.php?timestamp=16864…` (whatever the current timestamp is) and `https://tools.bytehamster.com/podcast/redirect2.php` contains `itunes:new-feed-url` in markup which contains the new feed URL (see also [this](https://podcasters.apple.com/support/837-change-the-rss-feed-url)).

**Reply by SamTV12345** [reply]
Weird Postman doesnt seem to do a redirect. It only returns a 200 on the base path. In the browser it indeed works. So in the browser I get redirected to the timestamp page but it doesn't contain the new feed url. How should my podcast server get that it needs to contact redirect2.php to retrieve the new podcasts?

From my understanding the workflow should be 
if redirectDetected
do 
   check itunes:new-feed-url
   replaceOldFeedWithNewFeed()

**Reply by alexanderadam** [reply]
@SamTV12345 I think these are just two different cases.
From my naïve point of view it should be something like

```ruby
# in the beginning it's feed_url = "https://tools.bytehamster.com/podcast/redirect.php"

# this is the first check on HTTP level
if [301, 308].include? response.code
  old_feed_urls << feed_url unless old_feed_urls.include? feed_url
  feed_url = response.url # https://tools.bytehamster.com/podcast/redirect.php?timestamp=16864…
end

# later when you do the parsing there could be another type of update by the data that you might find in the feed
if feed_xml.include? "itunes:new-feed-url"
  old_feed_urls << feed_url unless old_feed_urls.include? feed_url
  feed_url = feed_xml["itunes:new-feed-url"] # https://tools.bytehamster.com/podcast/redirect.php?timestamp=16864…
end
```

I hope that this Ruby pseudo code makes sense to you :wink: 
But I think the point is that the feed update can happen in two places (HTTP and XML).

**Reply by SamTV12345** [reply]
Thank you. I was a bit lost and though this was a sequence. That I get redirected from one page to another one and there is a "itunes:new-feed-url" denoted 😀 . Do I also have to update the feed items i.e. the podcast episodes? In the example the enclosure url stays the same. But I guess if I change e.g. the Podcast provider I also need to migrate the podcast episodes.

**Reply by alexanderadam** [reply]
@SamTV12345 

> Do I also have to update the feed items i.e. the podcast episodes?

I guess you'd need to do anyway (even if the URL didn't change) in case metadata (i.e. updated title) or url changes.
They should be recognizable via `guid` I guess.

**PS:** shouldn't we continue this conversation at [the dedicated issue](https://github.com/SamTV12345/PodFetch/issues/175)? Because here we are still at the AntennaPod sync question :wink: I updated the initial post with the updated data so that no one has to search in unrelated GitHub discussion feeds :wink:
